### PR TITLE
Thruster Control

### PR DESCRIPTION
**Pull request by Wesley-Fisher**
In Progress

### CODE REVIEW

**Review by amaarquadri — COMMENTED** [review]
In terms of computing the matrix that converts a wrench into the 6 motor thrusts, I think it would be better to read from a config file instead of having the xyzrpy values in the actual python file. That way we can have a single python script that can do all the computations, and we can easily swap out the config files or make updates.

File: catkin_ws/src/thruster_control/src/thruster_control/movement_command_collector.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+class MovementCommandCollector:
+    def __init__(self):
+        # Set up command sources
+        # First added is highest priority
+        self.sources = []
+        self.sources.append(CommandSubscriber("movement_command"))
+        self.sources.append(CommandSubscriber("depth_command"))
+
+    def get_recent_thrusts(self, drop=0):
+        w = self.get_empty_wrench()
```
Comment (Waterloo-Aquadrone) [inline catkin_ws/src/thruster_control/src/thruster_control/movement_command_collector.py]: rename

File: catkin_ws/src/thruster_control/src/thruster_control/movement_command_collector.py [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+        cmd, dt = source.get_cmd()
+        cmd = self.wrench_to_np_array(cmd)
+        if dt > self.cmd_timeout:
+            cmd = cmd * 0.0
+        return cmd
+
+    def get_empty_wrench(self):
+        return self.wrench_to_np_array(Wrench())
+    
+    def wrench_to_np_array(self, wrench):
```
Comment (Waterloo-Aquadrone) [inline catkin_ws/src/thruster_control/src/thruster_control/movement_command_collector.py:57]: potentially use a shared utility

File: catkin_ws/src/thruster_control/src/thruster_control/movement_command_collector.py [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    def get_cmd(self):
+        return self.cmd, (rospy.Time() - self.time).to_sec()
+
+
+class MovementCommandCollector:
+    def __init__(self):
+        # Set up command sources
+        # First added is highest priority
+        self.sources = []
+        self.sources.append(CommandSubscriber("movement_command"))
```
Comment (Waterloo-Aquadrone) [inline catkin_ws/src/thruster_control/src/thruster_control/movement_command_collector.py]: flip priority

File: catkin_ws/src/thruster_control/src/thruster_control/movement_command_collector.py [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+class CommandSubscriber:
+    # Should listen to a wrench command from some source topic
+    # Track the latest command, as well as the last time it was received
+
+    def __init__(self, topic):
+        self.cmd = Wrench()
+        self.time = rospy.Time()
+        self.sub = rospy.Subscriber(topic, Wrench, self.callback)
+
+        self.cmd_timeout = 0.5
```
Comment (Waterloo-Aquadrone) [inline catkin_ws/src/thruster_control/src/thruster_control/movement_command_collector.py]: move to MovementCommandCollector

File: catkin_ws/src/thruster_control/src/thruster_control/thruster_collection_manager.py [hunk 5]
```diff
@@ -0,0 +1,14 @@
+import rospy
+from aquadrone_msgs.msg import MotorControls
+
+class ThrusterCollectionManager:
+    num_thrusters = 6
```
Comment (Waterloo-Aquadrone) [inline catkin_ws/src/thruster_control/src/thruster_control/thruster_collection_manager.py]: use thrusters.size

File: catkin_ws/src/thruster_control/src/thruster_control/configurations/v1_configuration.py [hunk 6]
```diff
@@ -0,0 +1,26 @@
+import numpy as np
+import math
+
+import geometry_helper as gh
+
+def get_thrusts_to_wrench():
```
Comment (Waterloo-Aquadrone) [inline catkin_ws/src/thruster_control/src/thruster_control/configurations/v1_configuration.py]: TODO: read from file

File: catkin_ws/src/thruster_control/src/thruster_control/configurations/v1_configuration.py [hunk 7]
```diff
@@ -0,0 +1,26 @@
+import numpy as np
```
Comment (Waterloo-Aquadrone) [inline catkin_ws/src/thruster_control/src/thruster_control/configurations/v1_configuration.py:1]: TODO: make this a class

File: catkin_ws/src/thruster_control/src/thruster_control/configurations/v1_configuration.py (hunk 6 above)
Comment (Wesley-Fisher) [reply catkin_ws/src/thruster_control/src/thruster_control/configurations/v1_configuration.py]: Have marked for future project

**Review by ghost — APPROVED** [review]

**Pull request merged**